crates/deltalake-core/src/operations/write.rs

```diff
@@ -314,6 +327,14 @@ pub(crate) async fn write_execution_plan(
     };
 
     let checker = DeltaDataChecker::new(snapshot);
+    let checker = match predicate {
+        Some(pred) => {
+            // TODO: get the name of the outer-most column? `*` will also work but would be slower
+            let chk = DeltaConstraint::new("*", &fmt_expr_to_sql(&pred)?);
```

Comment:
Why is the predicate added as a delta constraint?

Reply:
this is to check that incoming data also conforms to the predicate (this is how the Spark Delta works). I could create a separate checker for it but this checker is only used in the context of this function so I thought integrating with existing checks could be another option. Happy to separate of course which would actually let me throw a more meaningful error in case of predicate check failure (at the moment it's only the generic check error message)

Reply:
Perhaps we should support this type of behaviour as well, either with the same configuration or a delta-rs table setting rust.delta.replaceWhere.constraintCheck.enabled ?

Good point but maybe deserves a separate issue? Delta is using `REPLACEWHERE_CONSTRAINT_CHECK_ENABLED` but default value is `true` so maybe add it in a separate PR?

Reply:
Check is enabled by default (I came across the behavior incidentally when I tried writing to a new table with no properties set and it didn't let me). Setting is here: 
https://github.com/delta-io/delta/blob/9f9993db8a51fd83168186bba24041feffdbb566/spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala#L753

Reply:
Ah, then it would make sense to also set this by default when we create a new table with the create api. And handle the cases where in some tables this setting is missing, which I guess means the same as if it was set to False?

Reply:
I disagree @ion-elgreco. Default value in Delta is `true` so there is no need to set it by default on create, and only set it if the caller has provided an explicit value. Well, it actually makes no harm to always setting it on create but

> And handle the cases where in some tables this setting is missing, which I guess means the same as if it was set to False?

Is not correct. In delta, if the setting is missing then the value is assumed `true` so we should adopt the same logic, right?

Reply:
which also means that if I don't implement that table setting for this PR then I'll be matching the behavior of the Delta by default

Reply:
Also, I presume we need to use the full config value as `spark.databricks.delta.replaceWhere.constraintCheck.enabled` if Delta is using that, right? (i've not touched the table properties yet)

Reply:
Actually, this config does not look like a table property but only an engine configuration, not sure how we are supposed to handle it: https://github.com/delta-io/delta/blob/9f9993db8a51fd83168186bba24041feffdbb566/docs/source/delta-batch.md?plain=1#L714C12-L714C12

Reply:
@roeap  or @rtyler, I don't recall who exactly, mentioned we should support these session settings as well. We could simply extend this in a future PR where we parse rust.delta environment variables to check which mode of predicate overwrite needs to be executed.